libpromises/eval_context.c

```diff
@@ -332,6 +346,19 @@ static StackFrame *LastStackFrameByType(const EvalContext *ctx, StackFrameType t
     return NULL;
 }
 
+static StackFrame *LastStackFrameByNonNullEvent(const EvalContext *ctx)
+{
+    for (size_t i = 0; i < SeqLength(ctx->stack); i++)
```

Comment:
To avoid computing the length more than once
```suggestion
    const size_t length = SeqLength(ctx->stack);
    for (size_t i = 0; i < length; i++)
```